### PR TITLE
Handle NuGet zero quirks by preserving version as intended

### PR DESCRIPTION
**Pull request by michaelnoonan**
NuGet introduced the concept of normalizing versions to help disambiguate packages that according to SemVer 2.0 are exactly the same version. For example `1.0.0.0` is semantically equivalent to `1.0.0`. And `2016.09.02` is semantically equivalent to 2016.9.2`. This is a great idea to avoid ambiguity in when comparing versions - but it doesn't help when you need to bridge outside the NuGet ecosystem.

For example: this behavior can break build scripts that expect the resulting package filename to match the version **exactly**, not **semantically**. The only workaround is to somehow use `SemanticVersion.Parse("2016.09.02").Normalize()` from within your scripts.

I started down the path of letting NuGet.exe do whatever it wanted, and then rename the resulting file - but Octopus prefers the NuGet metadata over the filename anyhow - so we need to align both the filename and the metadata.

Hence I've gone the whole hog and forked NuGet so we can use our own logic: https://github.com/OctopusDeploy/NuGet.Client/pull/1

Fixes https://github.com/OctopusDeploy/Issues/issues/2898
Relates to https://github.com/OctopusDeploy/Issues/issues/2809

### CODE REVIEW

**Comment by michaelnoonan** [comment]
I've updated this to use a `FileSystemWatcher` instead of solely parsing the log files. I think this feels like a reasonably solid solution and the tests are passing reliably.

**Comment by michaelnoonan** [comment]
Darn, taking packages built like this through to Octopus leads to more complexity:

1. The package is accepted by Octopus!
2. The file is written with the normalized filename (bummer - but makes sense)
3. ARC creates releases based on the normalized version (poop)
4. Manually created releases using the version from a donor package use the normalized version (poop)
5. Releases created by `octo.exe create-release ... --packageVersion=2016.01.01.0` may still work OK (cool)

It looks like we may need to be more invasive and fork `NuGet.exe` so it works like it used to...

File: readme.md [hunk 1]
```diff
@@ -92,6 +92,16 @@ When you use OctoPack, the NuGet package version number will come from (in order
...
 [You can see the logic here](https://github.com/OctopusDeploy/OctoPack/blob/master/source/OctoPack.Tasks/GetAssemblyVersionInfo.cs) 
 
+### Version Numbers are preserved as-is
+
+NuGet 3 started removing leading zeros and the fourth digit if it is zero. These are affectionately known as "NuGet zero quirks" and can be surprising when working with tooling outside the NuGet ecosystem. We have made a choice to preserve the version as-is when working with Octopus tooling to create packages of any kind. Learn more about [versioning in Octopus Deploy](http://docs.octopusdeploy.com/display/OD/Versioning+in+Octopus+Deploy).
+
+To make this work for NuGet packages we have forked NuGet
+
+The fork of NuGet 3 available here: https://github.com/OctopusDeploy/NuGet.Client
+The build is available here: http://build.octopushq.com/project.html?projectId=OctopusDeploy_NuGet&tab=projectOverview
```
Comment (robpearson) [inline readme.md]: Do we want this shared publicly?  I'm guessing it's read only.

File: tools/NuGet.exe.readme.txt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+This is the custom version of NuGet used by OctoPack.
+
+The primary customization is to preserve version strings as "what you see is what you get" for both the filename and metadata.
+
+This is a fork of NuGet 3 available here: https://github.com/OctopusDeploy/NuGet.Client
+The build is available here: http://build.octopushq.com/project.html?projectId=OctopusDeploy_NuGet&tab=projectOverview
```
Comment (robpearson) [inline tools/NuGet.exe.readme.txt]: Same thing here.  Include it here too?!?

Comment (michaelnoonan) [inline tools/NuGet.exe.readme.txt]: I put it there for convenience, plus it's restricted to only people on our domain.

**Pull request merged**